numpy/_core/src/multiarray/iterators.c

```diff
@@ -857,201 +744,170 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
+            goto finish;
         }
-        if (istrue) {
-            retval = PyArray_Pack(
-                    PyArray_DESCR(self->ao), self->dataptr, val);
+        ret = 0;
+        if (indices[0].value) {
+            ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
         }
         goto finish;
     }
 
-    if (PySequence_Check(ind) || PySlice_Check(ind)) {
-        goto skip;
-    }
-    start = PyArray_PyIntAsIntp(ind);
-    if (error_converting(start)) {
-        PyErr_Clear();
-    }
-    else {
-        if (check_and_adjust_index(&start, self->size, -1, NULL) < 0) {
+    PyArray_ITER_RESET(self);
+
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
             goto finish;
         }
-        PyArray_ITER_GOTO1D(self, start);
-        retval = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
+        PyArray_ITER_GOTO1D(self, indices[0].value);
+        ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
         PyArray_ITER_RESET(self);
-        if (retval < 0) {
+        if (ret < 0) {
             PyErr_SetString(PyExc_ValueError,
                             "Error setting single item of array.");
         }
         goto finish;
     }
 
- skip:
-    Py_INCREF(type);
-    arrval = (PyArrayObject *)PyArray_FromAny(val, type, 0, 0,
+    Py_INCREF(dtype);
+    PyArrayObject *arrval = (PyArrayObject *)PyArray_FromAny(val, dtype, 0, 0,
                                               NPY_ARRAY_FORCECAST, NULL);
     if (arrval == NULL) {
-        return -1;
+        goto finish;
     }
-    val_it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)arrval);
+    PyArrayIterObject *val_it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)arrval);
     if (val_it == NULL) {
         goto finish;
     }
     if (val_it->size == 0) {
-        retval = 0;
+        ret = 0;
         goto finish;
     }
 
     /* set up cast to handle single-element copies into arrval */
     NPY_ARRAYMETHOD_FLAGS transfer_flags = 0;
     npy_intp one = 1;
-    int itemsize = type->elsize;
     /* We can assume the newly allocated array is aligned */
     int is_aligned = IsUintAligned(self->ao);
     if (PyArray_GetDTypeTransferFunction(
-                is_aligned, itemsize, itemsize, type, type, 0,
+                is_aligned, dtype_size, dtype_size, dtype, dtype, 0,
                 &cast_info, &transfer_flags) < 0) {
         goto finish;
     }
 
-    /* Check Slice */
-    if (PySlice_Check(ind)) {
-        start = parse_index_entry(ind, &step_size, &n_steps, self->size, 0, 0);
-        if (start == -1) {
-            goto finish;
-        }
-        if (n_steps == ELLIPSIS_INDEX || n_steps == NEWAXIS_INDEX) {
-            PyErr_SetString(PyExc_IndexError,
-                            "cannot use Ellipsis or newaxes here");
+    if (index_type & HAS_SLICE) {
```

Comment:
```suggestion
    if (index_type == HAS_SLICE) {
```
let's change it here too.